### PR TITLE
2022-06-04: vectorization vs compilation

### PR DESCRIPTION
**Pull request by IcePigZDB**
Signed-off-by: IcePigZDB <icepigzdb@gmail.com>

### Summary of talk

Date: 2022-06-02

Topic: vectorization vs compilation

Keywords: Execution Engine, Volcano, Vectorization, Compilation

### Format check List

- [x] Conforming to English specifications especially in `README.md`.
- Capitalize the first letter.
- Use English punctuation with appropriate space.
- ...
- [x] Remember to add links to reference papers.
- [x] Make sure the link is accessible.
- Use dash `-` to connect filename rather than space ` `, the latter will casue dead link.
- [x] Make sure latex formulas display normally, more can be found in [#27](https://github.com/CDDSCLab/Weekly-Group-Meeting-Paper-List/issues/27).

### CODE REVIEW

File: meeting-summary/2022-06-02-Vectorization-vs-Compilation.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+## Reference
+
+[01] - [Boncz P A, Zukowski M, Nes N. MonetDB/X100: Hyper-Pipelining Query Execution[C]//Cidr. 2005, 5: 225-237.](http://cidrdb.org/cidr2005/papers/P19.pdf)\
+[02] - [Polychroniou O, Raghavan A, Ross K A. Rethinking SIMD vectorization for in-memory databases[C]//Proceedings of the 2015 ACM SIGMOD International Conference on Management of Data. 2015: 1493-1508.](http://www.cs.columbia.edu/~orestis/sigmod15.pdf)\
+[03] - [Neumann T. Efficiently compiling efficient query plans for modern hardware[J]. Proceedings of the VLDB Endowment, 2011, 4(9): 539-550.](https://www.vldb.org/pvldb/vol4/p539-neumann.pdf)\
+[04] - [Kohn A, Leis V, Neumann T. Adaptive execution of compiled queries[C]//2018 IEEE 34th International Conference on Data Engineering (ICDE). IEEE, 2018: 197-208.](https://db.in.tum.de/~leis/papers/adaptiveexecution.pdf)\
+[05] - [Kersten T, Leis V, Kemper A, et al. Everything you always wanted to know about compiled and vectorized queries but were afraid to ask[J]. Proceedings of the VLDB Endowment, 2018, 11(13): 2209-2222.](https://www.vldb.org/pvldb/vol11/p2209-kersten.pdf)\
+[05] – 参考博文: https://www.one-tab.com/page/LmMkDhRhRWKUxU2bRRJgQA\
```
Comment (ehds) [inline meeting-summary/2022-06-02-Vectorization-vs-Compilation.md:181]: [06],
new line

**Review by ehds — CHANGES_REQUESTED** [review]
rm a.cpp

File: meeting-summary/2022-06-02-Vectorization-vs-Compilation.md (hunk 1 above)
Comment (IcePigZDB) [reply meeting-summary/2022-06-02-Vectorization-vs-Compilation.md]: All done, ptal.

File: meeting-summary/2022-06-02-Vectorization-vs-Compilation.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+		case t0: return DoNothing
+		case t1: return Unoptimized
+		case t2: return Optimized
+```
+
+## 4 Vectorization vs Compilation [05]
+
+MonetDB(VectorWise) 和 Hyper 的作者 以及 CMU的 Andy Pavlo 等为了对比这两种具有共同目标而方向又截然不同的 Execution 方案的优缺点及适用场景。因为这些系统有很多独有的特殊优化，他们实现了 Tectorwise(VectorWise) 和 Typer(Hyper) 这两个原型，除了 execution 架构上，其余实现上尽可能相同，如 operator 的算法等。
+
+对 Single-Threaded Performance, SIMD, INTRA-QUERY PARALLELIZATION, HARDWARE, Compilation Time 等方面进行了实验对比。得出了一下结论：
```
Comment (ehds) [inline meeting-summary/2022-06-02-Vectorization-vs-Compilation.md:161]: `一下结论` -> `以下结论`

File: meeting-summary/2022-06-02-Vectorization-vs-Compilation.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+上面的 pipeline 和 pipeline breaker 指导了如何切分执行计划树。论文中对每个要 compilation 的 operator 定义了 `produce()` & `consume(attributes,source)` 两个需要实现的 abstraction 从而方便拓展，使得实现更加优雅。在 compilation 时使用深度优先遍历的顺序遍历执行计划树，每个算子递归调用它的 `produce` 接口进而调用其子算子的 `produce`, 计划树最底层是 scan 算子，其没有 `consume`接口，其`produce`接口会生成 scan 逻辑的代码，然后会调用父亲算子的 `consume` 生成他们的处理逻辑代码，完成后父亲算子会递归调用其父亲节点的 `consume` 直到没有上层节点。如此往复，即可将执行计划树转换成高级语言或者汇编语言，完成代码生成的全过程。
+
+这篇博客[Coding Husky](https://ericfu.me/code-gen-of-query/#more) 有个直观的[gif](https://ericfu.me/images/2019/03/code-gen-demo-animated.gif)。
+
+### 3.3 Adaptive Compilation (Hyper adaptive compilation 2018 [04])
+
+对于执行时间较长的 AP 负载，代码生成的 compilation time 相比于可能的性能提升来说往往时比较小的，是可以接受的，但是对于本身执行时间较短的 TP 负载，compilation 时间可能会比执行时间还长。
+<div  align="center">  <img src="imgs/20220602-three-mode.png"/> </div>
+争对这个问题，Hyper 实现了一种 adaptive compilation，根据不同负载选择不同的执行模式。为此 Hyper 首先实现了基于 vm 的 bytecode 解释执行模式，这种模式需要的编译时间很短，是所有 query 的默认执行模式。Hyper 使用 morsel-driven 的并行执行框架将 pipeline 按照输入切分，分成不同小任务以方便并行和模式切换。Hyper 会对 pipeline 多进程的执行平均速率，剩余 tuple 数量等进行监控，定时对不同模式完成盖 pipeline 剩余任务所需时间进行估计，选择所需最小时间的模式并在 thread 运行新的 morsel 时转换模式，如从 bytecode 解释运行转换为 compilation 模式。 注意 compilation 过程需要时间，此期间进程仍然可以以 bytecode 的模式处理。
```
Comment (ehds) [inline meeting-summary/2022-06-02-Vectorization-vs-Compilation.md:138]: `完成盖 pipeline`  means what?

**Review by ehds — APPROVED** [review]

**Comment by ehds** [comment]
LGTM

**Pull request merged**